openhtf/conf.py

```diff
@@ -294,334 +353,110 @@ def LoadFromDict(self, dictionary, force_reload=False):
-      _LOG.warning('Keyword arg %s not set from configuration, but is a '
-                   'configuration key', bad_name)
-
-    # Set positional args from configuration values.
-    config_args = {name: config[name] for name in arg_names if name in config}
-
-    for overridden in set(kwargs) & set(config_args):
-      _LOG.warning('Overriding provided kwarg %s=%s with value %s from '
-                   'configuration', overridden, kwargs[overridden],
-                   config_args[overridden])
-    kwargs.update(config_args)
-    _LOG.info('Invoking %s with %s', method.__name__, kwargs)
-    return method(**kwargs)
-
-  # We have to check for a 'self' parameter explicitly because Python doesn't
-  # pass it as a keyword arg, it passes it as the first positional arg.
-  if 'self' == argspec.args[0]:
-    @functools.wraps(method)
-    def SelfWrapper(self, **kwargs):  # pylint: disable=invalid-name,missing-docstring
-      kwargs['self'] = self
-      return method_wrapper(**kwargs)
-    return SelfWrapper
-  return method_wrapper
-
-
-# pylint: disable=invalid-name
-Declare = Config().model.Declare
-Load = Config().model.Load
-LoadMissingFromDict = Config().model.LoadMissingFromDict
-LoadFromDict = Config().model.LoadFromDict
-Reset = Config().model.Reset
-
-# Everywhere that uses configuration uses this, so we just declare it here.
-Declare('station_id', 'The name of this tester')
+    argspec = self._modules['inspect'].getargspec(method)
+
+    # Index in argspec.args of the first keyword argument.  This index is a
+    # negative number if there are any kwargs, or 0 if there are no kwargs.
+    keyword_arg_index = -1 * len(argspec.defaults or [])
+    arg_names = argspec.args[:keyword_arg_index or None]
+    kwarg_names = argspec.args[len(arg_names):]
+
+    # Create the actual method wrapper, all we do is update kwargs.  Note we
+    # don't pass any *args through because there can't be any - we've filled
+    # them all in with values from the configuration.  Any positional args that
+    # are missing from the configuration *must* be explicitly specified as
+    # kwargs.
+    @self._functools.wraps(method)
+    def method_wrapper(**kwargs):
+      """Wrapper that pulls values from openhtf.conf."""
+      # Check for keyword args with names that are in the config so we can warn.
+      for kwarg in kwarg_names:
+        if kwarg in self:
+          self._logger.warning('Keyword arg %s not set from configuration, but '
+                               'is a configuration key', kwarg)
+
+      # Set positional args from configuration values.
+      config_args = {name: self[name] for name in arg_names if name in self}
+
+      for overridden in set(kwargs) & set(config_args):
+        self._logger.warning('Overriding configuration value for kwarg %s (%s) '
+                             'with provided kwarg value: %s', overridden,
+                             self[overridden], kwargs[overridden])
+        del config_args[overridden]
```

Comment:
if you flipped the update method, you could elide this del, and it would make the operation clearer (configuration keys, except for what gets passed in)